src/kernel/arch/x86/gdt.zig

```diff
@@ -0,0 +1,293 @@
+    ecx: u32,
+
+    /// 32 bit data register.
+    edx: u32,
+
+    /// 32 bit base register.
+    ebx: u32,
+
+    /// 32 bit stack pointer register.
+    esp: u32,
+
+    /// 32 bit base pointer register.
+    ebp: u32,
+
+    /// 32 bit source register.
+    esi: u32,
+
+    /// 32 bit destination register.
+    edi: u32,
+
+    /// The extra segment.
+    es: u32,
+
+    /// The code segment.
+    cs: u32,
+
+    /// The stack segment.
+    ss: u32,
+
+    /// The data segment.
+    ds: u32,
+
+    /// A extra segment FS.
+    fs: u32,
+
+    /// A extra segment GS.
+    gs: u32,
+
+    /// The local descriptor table register.
+    ldtr: u32,
+
+    /// ?
+    trap: u16,
+
+    /// A pointer to a I/O port bitmap for the current task which specifies individual ports the program should have access to.
+    io_permissions_base_offset: u16,
+};
+
+///
+/// Make a GDT entry.
+///
+/// Arguments:
+///     IN access: u8 - The access bits for the descriptor.
+///     IN flags: u4  - The flag bits for the descriptor.
+///
+/// Return:
+///     A new GDT entry with the give access and flag bits set with the base at 0x00000000 and limit at 0xFFFFF.
+///
+fn makeEntry(base: u32, limit: u20, access: u8, flags: u4) GdtEntry {
+    return GdtEntry {
+        .limit_low = @truncate(u16, limit),
+        .base_low = @truncate(u24, base),
+        .access = access,
+        .limit_high = @truncate(u4, limit >> 16),
+        .flags = flags,
+        .base_high = @truncate(u8, base >> 24),
+    };
+}
+
+/// The GDT entry table of NUMBER_OF_ENTRIES entries.
+var gdt_entries: [NUMBER_OF_ENTRIES]GdtEntry = []GdtEntry {
+    // Null descriptor
+    makeEntry(0, 0, 0, 0),
+
+    // Kernel Code
+    makeEntry(0, 0xFFFFF, KERNEL_SEGMENT | CODE_SEGMENT, IS_32_BIT | IS_LIMIT_4K_BIT),
+
+    // Kernel Data
+    makeEntry(0, 0xFFFFF, KERNEL_SEGMENT | DATA_SEGMENT, IS_32_BIT | IS_LIMIT_4K_BIT),
+
+    // User Code
+    makeEntry(0, 0xFFFFF, USER_SEGMENT | CODE_SEGMENT, IS_32_BIT | IS_LIMIT_4K_BIT),
+
+    // User Data
+    makeEntry(0, 0xFFFFF, USER_SEGMENT | DATA_SEGMENT, IS_32_BIT | IS_LIMIT_4K_BIT),
+
+    // Fill in TSS at runtime
+    makeEntry(0, 0, 0, 0),
+};
+
+/// The GDT pointer that the CPU is loaded with that contains the base address of the GDT and the size.
+const gdt_ptr: GdtPtr = GdtPtr {
+    .limit = TABLE_SIZE,
+    .base = &gdt_entries[0],
+};
+
+/// The task state segment entry.
+var tss: TtsEntry = TtsEntry {
+    .prev_tss = 0,
+    .esp0 = undefined,
+    .ss0 = KERNEL_DATA_OFFSET,
+    .esp1 = 0,
+    .ss1 = 0,
+    .esp2 = 0,
+    .ss2 = 0,
+    .cr3 = 0,
+    .eip = 0,
+    .eflags = 0,
+    .eax = 0,
+    .ecx = 0,
+    .edx = 0,
+    .ebx = 0,
+    .esp = 0,
+    .ebp = 0,
+    .esi = 0,
+    .edi = 0,
+    .es = 0,
+    .cs = 0,
+    .ss = 0,
+    .ds = 0,
+    .fs = 0,
+    .gs = 0,
+    .ldtr = 0,
+    .trap = 0,
+    .io_permissions_base_offset = @sizeOf(TtsEntry),
+};
+
+pub fn setKernelStack(esp0: u32) void {
```

Comment:
Perhaps `setTssStack` would be a more descriptive name?

Reply:
Sure